### PR TITLE
Resolve open batch boundary issues

### LINKED ISSUE #118: Batch constraints are a bit funny
The current text has two batch constraints:

```
*  `batch_size`: The batch size, i.e., the minimum number of reports that are
  aggregated into an output.
* `batch_window`: The window of time covered by a batch, i.e., the maximum
  interval between the oldest and newest report in a batch.
```

This seems like it could have the result that collection is impossible. Suppose that we have:

```
batch_size = 7200
batch_window = 3600
```

And reports come in at an average of 1/s. In this case it wouldn't be possible to satisfy these constraints because you would never be able to accumulate 7200 reports. More generally, it's not clear to me why ```batch_window``` should be a maximum rather than a minimum. How does that assist the helper/leader.

ISTM we should either have one constraint (probably ```batch_size```) or two but have them both be minimums. In either case, we probably want some way for the collector to learn when the limits are reached rather than having to poll.

### PR DESCRIPTION
**Pull request by cjpatton**
We have the following security/operational considerations, not all of
which can be addressed at the same time:

  - Enforcing privacy budgets for DP requires us to limit the number of
  batches the same report is used.
  - Hits requires the same report to be used in some number of batches.

To resolve these issues, this change adds a `max_report_lifetime`
parameter, which specifies the maximum number of times a report may be
aggregated into an output. This enables Hits while allowing PPM
protocols to enforce a privacy budget.

This change also seeks to clarify some ambiguities in how the batch
boundaries are defined, and why they're defined that way. Along the way,
it cleans up some protocol bits that got lost in the last refactor.

Closes #118 by doing nothing. Hopefully the new text makes the issue more
clear.

### CODE REVIEW

File: draft-ppm-protocol.md
Comment (tgeoghegan) [inline draft-ppm-protocol.md:1115]: This would also be a good place to discuss efficient means of tracking privacy budgets.

**Review by tgeoghegan — CHANGES_REQUESTED** [review]

File: draft-ppm-protocol.md
Comment (cjpatton) [reply draft-ppm-protocol.md:973]: Yes.

File: draft-ppm-protocol.md
Comment (cjpatton) [reply draft-ppm-protocol.md:983]: There is indeed an increase of storage requirements compared to the status quo. However, the status quo doesn't account for privacy budgets at all. Hence, I claim this is a cost we have to eat no matter what.

But how big is the cost? The key thing to notice is that the increase in storage requirements is linear in the number of batch intervals _that can still be queried_. This number depends on the minimum batch duration and the rate at which the collector makes requests. Crucially, the storage increase _does not depend on the number of reports_.

File: draft-ppm-protocol.md
Comment (cjpatton) [reply draft-ppm-protocol.md:984]: Interesting question. To put it another way: Is the max report lifetime too coarse-grained?

The interesting thing about Hits is that, without DP, _some amount of leakage_ is inevitable, but as you point out, how much is leaked depends on the distribution of inputs. (As an extreme example, suppose a large number of non-heavy-hitters share a common prefix.)

Vanilla Hits doesn't try to mitigate this at all, so we would set the max report lifetime to be the length of the input. (This is the maximum number of collect requests that would be needed to compute the heavy hitters.) The current proposal for mitigating this leakage is via DP. Interestingly, if the aggregators are adding the noise, then the distribution of the noise -- and therefore the privacy budget -- can only depend on the candidate prefixes of the previous rounds. I don't imagine it would, it seems reasonable to just pick a max report lifetime. But we don't want to preclude the possibility of someone doing something fancier.

Of course, nothing about the protocol precludes the aggregators from bailing out _before_ the max report lifetime is reached... what's important is that they agree on when to bail out.

Which leads me to this conclusion: the max report lifetime parameter seems to wear to hats:
1. It provides a coarse-grained way of enforcing a privacy budget. In particular, it should suffice for "simple" schemes in which the privacy budget is independent of the input distribution.
2. It provides a way for aggregators to know how long they have to keep around reports (or the output shares derived from them). Namely, if a report has reached its max lifetime, it's OK to throw it away.

This PR talks about (1.), but perhaps it also makes sense to talk about (2.).

File: draft-ppm-protocol.md
Comment (cjpatton) [reply draft-ppm-protocol.md:990]: You're right, this needs to be rephrased. What do you think about

> Using a report multiple times withing a single batch, or using the same report in multiple batches, is considered a privacy violation. To prevent such attacks...

File: draft-ppm-protocol.md
Comment (cjpatton) [reply draft-ppm-protocol.md:1115]: We already suggest that such a scheme exists:
>  each aggregator only needs to store the aggregate output share for each possible batch interval,
along with the number of times the output share was used in a batch.

I suggest we hold off on describing this until we've implemented it. This will give us a chance to validate any assumptions we're making here. If you'd like I can flag an OPEN ISSUE here.

File: draft-ppm-protocol.md
Comment (cjpatton) [reply draft-ppm-protocol.md:1354]: Yup, good call. Maybe this distinction would be more clear if we renamed "max_report_lifetime" to "max_batch_lifetime", since what we're really talking about is how many times a batch of reports can be queried.

File: draft-ppm-protocol.md
Comment (cjpatton) [reply draft-ppm-protocol.md:140]: Yeah, I can do that.

**Comment by cjpatton** [comment]
Rebased.

File: draft-ppm-protocol.md
Comment (tgeoghegan) [reply draft-ppm-protocol.md:994]: >Here I disagree. This operational parameter is a kind of contract between collectors and aggregators that stipulates how long data has to be available. Without it, aggregators would have to keep state indefinitely. I suppose we could do this in a different way, but we'll need to have something like this in the spec.

I'm not arguing for getting rid of this parameter, just for changing its name to better reflect what its units are (a number of queries rather than a duration). I would phrase it as "once a batch's `privacy_budget` is exhausted, aggregators MUST refuse further queries on the batch and MAY delete the inputs in the batch" (note that the MUST part is something observable by other parties whereas it's not meaningfully possible for an aggregator to prove to other participants that it has deleted some input).

But as I said, we should move forward with `max_batch_lifetime`.

**Review by tgeoghegan — APPROVED** [review]
A few more nits but basically LGTM once any occurrence of "batch window" is corrected.

File: draft-ppm-protocol.md
Comment (cjpatton) [reply draft-ppm-protocol.md]: Done.

**Comment by cjpatton** [comment]
Squashed.

**Pull request merged**
